src/api/tap.rs

```diff
@@ -0,0 +1,401 @@
+    pub fn matched(input_name: &str) -> Self {
+        Self::Notification(input_name.to_string(), TapNotification::Matched)
+    }
+
+    pub fn not_matched(input_name: &str) -> Self {
+        Self::Notification(input_name.to_string(), TapNotification::NotMatched)
+    }
+}
+
+/// A `TapSink` is used as an output channel for a topology component, and receives
+/// `Event`s. If these are of type `Event::LogEvent`, they are relayed to the tap client.
+pub struct TapSink {
+    tap_tx: TapSender,
+    component_name: String,
+    buffer: VecDeque<LogEvent>,
+}
+
+impl TapSink {
+    pub fn new(tap_tx: TapSender, component_name: String) -> Self {
+        Self {
+            tap_tx,
+            component_name,
+            // Pre-allocate space of 100 events, which matches the default `limit` typically
+            // provided to a tap subscription. If there's a higher log volume, this will block
+            // until the upstream event handler has processed the event. Generally, there should
+            // be little upstream pressure in the processing pipeline.
+            buffer: VecDeque::with_capacity(100),
+        }
+    }
+}
+
+impl Sink<Event> for TapSink {
+    type Error = ();
+
+    /// The sink is ready to accept events if buffer capacity hasn't been reached.
+    fn poll_ready(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.buffer.len() == self.buffer.capacity() {
+            Poll::Pending
+        } else {
+            Poll::Ready(Ok(()))
+        }
+    }
+
+    /// If the sink is ready, and the event is of type `LogEvent`, add to the buffer.
+    fn start_send(mut self: Pin<&mut Self>, item: Event) -> Result<(), Self::Error> {
+        // If we have an event, push it onto the buffer.
+        if let Event::Log(ev) = item {
+            self.buffer.push_back(ev);
+        }
+
+        Ok(())
+    }
+
+    /// Flushing means FIFO dequeuing. This is an O(1) operation on the `VecDeque` buffer.
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        let mut tap_tx = self.tap_tx.clone();
+
+        // Loop over the buffer events, pulling from the front. This will terminate when
+        // the buffer is empty.
+        while let Some(ev) = self.buffer.pop_front() {
+            // Attempt to send upstream. If the channel is closed, log and break. If it's
+            // full, return pending to reattempt later.
+            match tap_tx.try_send(TapPayload::Log(self.component_name.clone(), ev)) {
+                Err(tokio_mpsc::error::TrySendError::Closed(payload)) => {
+                    debug!(
+                        message = "Couldn't send log event.",
+                        payload = ?payload,
+                        component_name = ?self.component_name);
+
+                    break;
+                }
+                Err(tokio_mpsc::error::TrySendError::Full(_)) => return Poll::Pending,
```

Comment:
Again here, I don't think it's safe to return `Pending` if we don't want to be a source of backpressure.

Reply:
This now returns `Poll::Ready(Ok(()))`.